Fix heap corruption

krish2718 opened the pull request.

krish2718 committed:
[nrf toup] packet: Add Malloc checks

Check for allocation failure and log an error. The error aren't
propagated as it needs more changes, but at least we can identify that
allocation has failed.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
[nrf toup] packet: Fix heap corruption

While parsing the packet all TLV's are NULl terminated explicitly, so,
take in to account the NULL terminating byte while memory allocation.

Else, this overrides other memory causing hard to debug heap
corruptions.

Fixes SHEL-2754.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

bama-nordic commented on indigo_api.c:
```diff
@@ -322,15 +322,31 @@ void fill_wrapper_ack(struct packet_wrapper *wrapper, int seq, int status, char
 
     wrapper->tlv_num =  2;
     wrapper->tlv[0] = malloc(sizeof(struct tlv_hdr));
+    if (!wrapper->tlv[0]) {
```
Wondering if this whole "if (xyz) { }" part can be made a macro - may look easier on the eye and reusable throughout the codebase?

krish2718 replied:
Yes, the entire QT code can be heavily refactored, but as this is OSS code, we are trying to keep the changes to minimal for easier upmerge (if at all we do one in the future).

bama-nordic replied:
Amazing how buggy the original OSS code is!!

krish2718 replied:
and they aren't even accepting PRs :)

krish2718 replied:
and as I said in the commit logging, logging allocation failures is just a partial fix, all these functions return `void`, so, cannot propagate these errors, so, it might crash later, but at least we would know it's because of allocation failure.

krish2718 committed:
[nrf noup] zephyr: Fix POSIX failures

When using POSIX_API we need to define dummy prototype for compilation
to succeed.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>
